x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+                            boolean accepted = runAutomaton.run(decoded, 0, decoded.length);
+                            if (accepted) {
+                                return AcceptStatus.YES;
+                            }
+                            return AcceptStatus.NO;
+                        }
+                    };
+                }
+            };
+        }
+
+        @Override
+        public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[wildcard] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+
+            VersionFieldWildcardQuery query = new VersionFieldWildcardQuery(new Term(name(), value));
+            QueryParsers.setRewriteMethod(query, method);
+            return query;
+        }
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            String valueAsString;
+            if (value instanceof String) {
+                valueAsString = (String) value;
+            } else if (value instanceof BytesRef) {
+                // encoded string, need to re-encode
+                valueAsString = ((BytesRef) value).utf8ToString();
+            } else {
+                throw new IllegalArgumentException("Illegal value type: " + value.getClass() + ", value: " + value);
+            }
+            return encodeVersion(valueAsString).bytesRef;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new SortedSetOrdinalsIndexFieldData.Builder(name(), VersionScriptDocValues::new, CoreValuesSourceType.BYTES);
+        }
+
+        @Override
+        public Object valueForDisplay(Object value) {
+            if (value == null) {
+                return null;
+            }
+            return VERSION_DOCVALUE.format((BytesRef) value);
+        }
+
+        @Override
+        public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
+            if (format != null) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support custom formats");
+            }
+            if (timeZone != null) {
+                throw new IllegalArgumentException(
+                    "Field [" + name() + "] of type [" + typeName() + "] does not support custom time zones"
+                );
+            }
+            return VERSION_DOCVALUE;
+        }
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```

Comment:
I am wondering if this query still should be considering expensive if with  a binaryPoint query we can efficiently narrow down terms?

Reply:
I was also wondering since lowerTerm and upperTerm represent valid versions can we extract major, minor, patch fields  and use them for querying to further narrow down terms?

Reply:
lower and upper don't necessarily represent valid versions, we'd also be able to use non-semver versions that then sort above the valid ones. Major/Minor/Patch might be not available for them.  As for the expensive part, I'm happy to change, I'm not familiar enough with the cost of binaryPoint so I'd follow your judgement or if anyone else has an opinion on this.